### DISCUSSION TITLE
Archiver Identify tool with just config given throws an error

### DISCUSSION THREAD

**Discussion by DavinRK1** [thoughtspot/cs_tools#151] [General]
After recently getting CS tools downloaded I have been attempting to run some of the tools to test and learn with our lower environments. I have been trying to run the archiver tool in our QA environment with different options given. I today try to just give the config and only run the default settings in the tool and received this error. 
╭──────────────────────────────── Traceback (most recent call last) ─────────────────────────────────╮
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\archiver\ │
│ app.py:125 in identify                                                                             │
│                                                                                                    │
│   122 │   ]                                                                                        │
│   123 │                                                                                            │
│   124 │   with LiveTasks(tasks, console=rich_console) as tasks:                                    │
│ ❱ 125 │   │   with tasks["gather_ts_bi"]:                                                          │
│   126 │   │   │   ts_bi_rows = ts.search(                                                          │
│   127 │   │   │   │   f"[user action] != [user action].answer_unsaved [user action].{{null}} "     │
│   128 │   │   │   │   f"[answer book guid] != [answer book guid].{{null}} "                        │
│                                                                                                    │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\layout.py:115   │
│ in __exit__                                                                                        │
│                                                                                                    │
│   112 │   │   self.stop(error=bool(exc))                                                           │
│   113 │   │                                                                                        │
│   114 │   │   if exc is not None:                                                                  │
│ ❱ 115 │   │   │   raise exc                                                                        │
│   116                                                                                              │
│   117                                                                                              │
│   118 class LiveTasks(Live):                                                                       │
│                                                                                                    │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\archiver\ │
│ app.py:126 in identify                                                                             │
│                                                                                                    │
│   123 │                                                                                            │
│   124 │   with LiveTasks(tasks, console=rich_console) as tasks:                                    │
│   125 │   │   with tasks["gather_ts_bi"]:                                                          │
│ ❱ 126 │   │   │   ts_bi_rows = ts.search(                                                          │
│   127 │   │   │   │   f"[user action] != [user action].answer_unsaved [user action].{{null}} "     │
│   128 │   │   │   │   f"[answer book guid] != [answer book guid].{{null}} "                        │
│   129 │   │   │   │   f"[timestamp].'last {recent_activity} days' "                                │
│                                                                                                    │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\api\middlewares\sea │
│ rch.py:236 in __call__                                                                             │
│                                                                                                    │
│   233 │   │                                                                                        │
│   234 │   │   # Get the data types                                                                 │
│   235 │   │   column_name = "referenced_column_name" if use_logical_column_names else "name"       │
│ ❱ 236 │   │   data_types = {column[column_name]: column["data_type"] for column in d["columnDe     │
│   237 │   │                                                                                        │
│   238 │   │   # Remap column names based on the desired column state                               │
│   239 │   │   d["columnNames"] = list(data_types.keys())                                           │
╰────────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'columnDetails'

          ╭───────────────────── 😱  This is an unhandled error!  😱 ─────────────────────╮
          │                                                                               │
          │ If you encounter this message more than once, please help by letting us know! │
          │                                                                               │
          │           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose   │
          │                                                                               │
          ╰──────────── Run cs_tools logs report to send us your last error. ─────────────╯
          
          My input was: cs_tools tools archiver identify --config qa
          
          This config file worked when using the user-management rename tool which makes me believe this is not an error with the config file? 

Let me know if I can provide anymore information, thank you.

**Comment by boonhapus** [comment]
Hey @DavinRK1 , sorry about this one! I'll have it fixed by Monday.

**Reply by DavinRK1** [reply]
Great, thank you!

**Reply by DavinRK1** [reply]
Hi @boonhapus Just want to thank you again for the help on CS Tools in our org. I was on a call with TS support today to discuss more about CS Tools and how to use them, as we were going through some use cases for other tools we ran into this error with bulk-deleter single and the told me to post it to this thread.

 cs_tools tools bulk-deleter single --object-type PINBOARD_ANSWER_BOOK --object-guid a12b0ab9-7d76-4f65-95b6-3e253da74696 --config qa
Fetching latest CS Tools release version timed out after 0.33s
╭─ An update has been released! 🎉 ──────────╮
│                                            │
│ Out now, CS Tools version v1.5.8b8!        │
│                                            │
│ Check out the changes in the Release Notes │
│                                            │
│ Upgrade with cs_tools self update          │
╰────────────────────────────────────────────╯
[13:06:33] INFO     Attempting Basic authentication                                 thoughtspot.py:213
╭──────────────────────────────── Traceback (most recent call last) ─────────────────────────────────╮
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\bulk-dele │
│ ter\app.py:33 in single                                                                            │
│                                                                                                    │
│    30 │   Removes a specific object from ThoughtSpot.                                              │
│    31 │   """                                                                                      │
│    32 │   ts = ctx.obj.thoughtspot                                                                 │
│ ❱  33 │   data = work._validate_objects_exist(ts, data=[{"object_type": object_type, "object_g     │
│    34 │                                                                                            │
│    35 │   if not data:                                                                             │
│    36 │   │   log.info("[red]found no valid objects to delete")                                    │
│                                                                                                    │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\bulk-dele │
│ ter\work.py:30 in _validate_objects_exist                                                          │
│                                                                                                    │
│   27 │                                                                                             │
│   28 │   This is a ThoughtSpot API limitation.                                                     │
│   29 │   """                                                                                       │
│ ❱ 30 │   data = [types.DeleteRecord(**content) for content in data]                                │
│   31 │                                                                                             │
│   32 │   answers = [content for content in data if content.object_type == "QUESTION_ANSWER_BO      │
│   33 │   liveboards = [content for content in data if content.object_type == "PINBOARD_ANSWER      │
│ in __init__:7                                                                                      │
│                                                                                                    │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\bulk-dele │
│ ter\types.py:36 in __post_init__                                                                   │
│                                                                                                    │
│   33 │   status: str = ":cross_mark:"                                                              │
│   34 │                                                                                             │
│   35 │   def __post_init__(self):                                                                  │
│ ❱ 36 │   │   self.object_type = AcceptedObjectType[self.object_type].system_type                   │
│   37 │                                                                                             │
│   38 │   @property                                                                                 │
│   39 │   def values(self) -> tuple[str]:                                                           │
│                                                                                                    │
│ C:\Program Files\Python312\Lib\enum.py:814 in __getitem__                                          │
╰────────────────────────────────────────────────────────────────────────────────────────────────────╯
TypeError: unhashable type: 'AcceptedObjectType'


          ╭───────────────────── 😱  This is an unhandled error!  😱 ─────────────────────╮
          │                                                                               │
          │ If you encounter this message more than once, please help by letting us know! │
          │                                                                               │
          │           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose   │
          │                                                                               │
          ╰──────────── Run cs_tools logs report to send us your last error. ─────────────╯



Thank you!

**Reply by boonhapus** [reply]
Hey @DavinRK1 can you shoot me an email?

nicholas.cooper@thoughtspot.com

I'd prefer we don't go through the Support team to get help with the tools unless there's an actual issue with ThoughtSpot itself.

I'd also love to get to know what you're trying to achieve with the tools to better help you. 🙂

**Reply by DavinRK1** [reply]
Hey @boonhapus Thanks again for the new version via Email, sadly when trying to run dry-run on the archiver identify tool I received the same error as previously. Error below.

PS C:\Users\XFMJ7YY> cs_tools tools archiver identify --dry-run --config qa
Fetching latest CS Tools release version timed out after 0.33s
[15:46:12] INFO     Attempting Basic authentication                                                   thoughtspot.py:213

  Status    Started At    Duration (s)   Task
 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    ❌       19:46:13             1.66   Getting content usage and activity statistics
    🍿                                   Getting supporting metadata for content identification
    🍿                                   Getting existing content metadata
    🍿                                   Writing Archiver syncer
    🍿                                   Showing a sample of 25 items to tag with INACTIVE
    🍿                                   Confirmation prompt
    🍿                                   Tagging content in ThoughtSpot



╭───────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────╮
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\archiver\app.py:125 in      │
│ identify                                                                                                             │
│                                                                                                                      │
│   122 │   ]                                                                                                          │
│   123 │                                                                                                              │
│   124 │   with LiveTasks(tasks, console=rich_console) as tasks:                                                      │
│ ❱ 125 │   │   with tasks["gather_ts_bi"]:                                                                            │
│   126 │   │   │   ts_bi_rows = ts.search(                                                                            │
│   127 │   │   │   │   f"[user action] != [user action].answer_unsaved [user action].{{null}} "                       │
│   128 │   │   │   │   f"[answer book guid] != [answer book guid].{{null}} "                                          │
│                                                                                                                      │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\layout.py:115 in __exit__         │
│                                                                                                                      │
│   112 │   │   self.stop(error=bool(exc))                                                                             │
│   113 │   │                                                                                                          │
│   114 │   │   if exc is not None:                                                                                    │
│ ❱ 115 │   │   │   raise exc                                                                                          │
│   116                                                                                                                │
│   117                                                                                                                │
│   118 class LiveTasks(Live):                                                                                         │
│                                                                                                                      │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\archiver\app.py:126 in      │
│ identify                                                                                                             │
│                                                                                                                      │
│   123 │                                                                                                              │
│   124 │   with LiveTasks(tasks, console=rich_console) as tasks:                                                      │
│   125 │   │   with tasks["gather_ts_bi"]:                                                                            │
│ ❱ 126 │   │   │   ts_bi_rows = ts.search(                                                                            │
│   127 │   │   │   │   f"[user action] != [user action].answer_unsaved [user action].{{null}} "                       │
│   128 │   │   │   │   f"[answer book guid] != [answer book guid].{{null}} "                                          │
│   129 │   │   │   │   f"[timestamp].'last {recent_activity} days' "                                                  │
│                                                                                                                      │
│ C:\Users\XFMJ7YY\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\api\middlewares\search.py:236 in      │
│ __call__                                                                                                             │
│                                                                                                                      │
│   233 │   │                                                                                                          │
│   234 │   │   # Get the data types                                                                                   │
│   235 │   │   column_name = "referenced_column_name" if use_logical_column_names else "name"                         │
│ ❱ 236 │   │   data_types = {column[column_name]: column["data_type"] for column in d["columnDe                       │
│   237 │   │                                                                                                          │
│   238 │   │   # Remap column names based on the desired column state                                                 │
│   239 │   │   d["columnNames"] = list(data_types.keys())                                                             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'columnDetails'


                   ╭───────────────────── 🤓  This is an unhandled error!  🤓 ─────────────────────╮
                   │                                                                               │
                   │ If you encounter this message more than once, please help by letting us know! │
                   │                                                                               │
                   │           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose   │
                   │                                                                               │
                   ╰──────────── Run cs_tools logs report to send us your last error. ─────────────╯

PS C:\Users\XFMJ7YY> cs_tools --version
Fetching latest CS Tools release version timed out after 0.33s


╭─ CS Tools ─╮
│            │
│   1.5.8    │
│            │
╰────────────╯

Thanks!
